### PR TITLE
Limit depth of recursion for join of types sharing structure

### PR DESCRIPTION
**Pull request by lthls**
This PR is aimed at limiting the amount of duplication one can get when joining two structures that were created in a similar way from different sets of input values, containing several references to these inputs.

For instance:
```ocaml
let  make_struct v =
  (v, (v, v))

let v1 = (0, 1)
let v2 = (2, 3)
let r = if foo then make_struct v1 else make_struct v2
```

Setting the new `join_depth` parameter to 0 should give `r` a type looking like `(T, T)`, with depth 1 it should get `((T, T), (T, T))`, with depth 2 it should get `(({0, 2}, {1, 3}), ((T, T), (T, T)))` and with depth 3 or more it should get the full `(({0, 2}, {1, 3}), (({0, 2}, {1, 3}), ({0, 2}, {1, 3})))` that is computed without this PR.

I still need to test that it works as expected, and I'm not sure yet what the default parameters should be.

### CODE REVIEW

**Comment by lthls** [comment]
It's getting a bit tricky to test this properly: in addition to the usual interference from other optimisations, I have to examine the intermediate typing environments (everything that goes through `make_suitable_for_environment` gets cleaned up, in a different way). Plus some of the equations that I would expect to be lost are re-introduced by the subkind information, which confused me for a while.

But speaking with @Gbury about this, we noticed that there might be an other "solution": turning the `-flambda2-expert-max-unboxing-depth` setting higher. Unboxing forces the creation of common aliases, so this should have the consequence of not triggering the problematic case where we join two aliases and cannot find a shared alias for them.
This might prove a better short-term fix than this PR.

**Comment by lukemaurer** [comment]
This seems to fix #472.

**Comment by lthls** [comment]
I've updated the default depth to 5, as suggested by @chambart.
I think this can be merged for now, and we'll keep working on a better solution when we find the time.

File: driver/flambda_backend_flags.ml [hunk 1]
```diff
@@ -216,13 +218,15 @@ module Flambda2 = struct
...
     cse_depth := 2;
+    join_depth := 2;
     join_points := false;
     unbox_along_intra_function_control_flow := true;
     Expert.fallback_inlining_heuristic := false;
     backend_cse_at_toplevel := false
 
   let o3_flags () =
     cse_depth := 2;
+    join_depth := 2;
```
Comment (mshinwell) [inline driver/flambda_backend_flags.ml:229]: Is this supposed to say 2?  The default is set to 5.

**Review by mshinwell — APPROVED** [review]

File: driver/flambda_backend_flags.ml (hunk 1 above)
Comment (lthls) [reply driver/flambda_backend_flags.ml:229]: Ah, that's an oversight. I don't remember why I added this line, so I'll remove it.

**Pull request merged**